module/zfs/vdev_indirect.c

```diff
@@ -1268,19 +1268,14 @@ vdev_indirect_read_all(zio_t *zio)
 		for (int i = 0; i < is->is_children; i++) {
 			indirect_child_t *ic = &is->is_child[i];
 
-			if (!vdev_readable(ic->ic_vdev))
-				continue;
-
 			/*
-			 * Note, we may read from a child whose DTL
-			 * indicates that the data may not be present here.
-			 * While this might result in a few i/os that will
-			 * likely return incorrect data, it simplifies the
-			 * code since we can treat scrub and resilver
-			 * identically.  (The incorrect data will be
-			 * detected and ignored when we verify the
-			 * checksum.)
+			 * In case of a resilver do not read from a child
+			 * whose DTL indicates it may be missing data.
 			 */
+			if (!vdev_readable(ic->ic_vdev) ||
+			    ((zio->io_flags & ZIO_FLAG_RESILVER) &&
+			    !vdev_dtl_empty(ic->ic_vdev, DTL_MISSING)))
+				continue;
```

Comment:
How about this.  Rather than skipping a child since it _may_ not contain the correct data I think it would be preferable to do the read but set a flag indicating the data may be incorrect.

Setting a flag gives us every opportunity to reconstruct the split block if it's at all possible which is inline with how all of the existing indirect vdev repair logic is intended to work.  Both raidz and draid roughly [work in the same way](https://github.com/openzfs/zfs/blob/master/module/zfs/vdev_raidz.c#L1791-L1799) for the reconstruction cases which are similar (`rc_error = ESTALE` is set). Then in `vdev_indirect_repair()` you can check the flag and suppress incrementing the checksum counter since we know it was a long shot to begin with.

To check the DTL you'll want to do something like this `vdev_dtl_contains(cvd, DTL_MISSING, zio->io_txg, 1)`.

Reply:
Thanks for the feedback, will do!